test/test_shape_bucketing.py

```diff
@@ -0,0 +1,308 @@
+        @kernel(
+            settings=Settings(
+                static_shapes=False,
+                shape_bucketing="zero_nonzero",
+                autotune_effort="none",
+            )
+        )
+        def pw_add(x: torch.Tensor, out: torch.Tensor) -> None:
+            for i in grid(x.size(0)):
+                for j in grid(x.size(1)):
+                    out[i, j] = x[i, j] + 1.0
+
+        device = torch.device("cuda", 0)
+        K = 16
+
+        x_col = torch.randn(K, 1, device=device, dtype=torch.float32)
+        y_col = torch.empty_like(x_col)
+        x_row = torch.randn(1, K, device=device, dtype=torch.float32)
+        y_row = torch.empty_like(x_row)
+
+        b = pw_add.bind((x_col, y_col))
+
+        # First call compiles once
+        pw_add(x_col, y_col)
+        torch.testing.assert_close(y_col, x_col + 1.0, rtol=1e-4, atol=1e-4)
+        self.assertEqual(len(b._compile_cache), 1)
+
+        # Flip which dim is 1; still correct; cache unchanged
+        pw_add(x_row, y_row)
+        torch.testing.assert_close(y_row, x_row + 1.0, rtol=1e-4, atol=1e-4)
+        self.assertEqual(len(b._compile_cache), 1)
+
+    @skipIfNotCUDA()
+    def test_zero_nonzero_codegen_identical_m1_vs_m2(self) -> None:
+        # Under zero_nonzero, M=1 vs M=2 should produce identical codegen
+        def pw_add_fn(x: torch.Tensor, out: torch.Tensor) -> None:
+            for i in grid(x.size(0)):
+                for j in grid(x.size(1)):
+                    out[i, j] = x[i, j] + 1.0
+
+        device = torch.device("cuda", 0)
+        K = 16
+        settings = Settings(
+            static_shapes=False, autotune_effort="none", shape_bucketing="zero_nonzero"
+        )
+
+        k1 = kernel(pw_add_fn, settings=settings)
+        k2 = kernel(pw_add_fn, settings=settings)
+
+        x1 = torch.randn(1, K, device=device, dtype=torch.float32)
+        y1 = torch.empty_like(x1)
+        b1 = k1.bind((x1, y1))
+        code1 = b1.to_triton_code()
+        self.assertExpectedJournal(code1)
+
+        x2 = torch.randn(2, K, device=device, dtype=torch.float32)
+        y2 = torch.empty_like(x2)
+        b2 = k2.bind((x2, y2))
+        code2 = b2.to_triton_code()
+        self.assertExpectedJournal(code2)
+
+        self.assertEqual(code1, code2)
+
+    @skipIfNotCUDA()
+    def test_zero_nonzero_runtime_correctness_varying_singleton_dim_3d(self) -> None:
+        # Compile at (1, K, K) then run across different 3D 1-ness patterns; must be correct and reuse a single compiled callable
+        @kernel(
+            settings=Settings(
+                static_shapes=False,
+                shape_bucketing="zero_nonzero",
+                autotune_effort="none",
+            )
+        )
+        def pw_add3d(x: torch.Tensor, out: torch.Tensor) -> None:
+            for i in grid(x.size(0)):
+                for j in grid(x.size(1)):
+                    for k in grid(x.size(2)):
+                        out[i, j, k] = x[i, j, k] + 1.0
```

Comment:
Why do all the tests use grid rather than tile?  This seems a bit odd.  
```suggestion
            for i, j, k in tile(x.size()):
                        out[i, j, k] = x[i, j, k] + 1.0
```

Reply:
more of a habit of using `grid`, coming from triton. using `tile` would make sense here.